lib/Dialect/Pipeline/Transforms/ScheduleLinearPipeline.cpp

```diff
@@ -0,0 +1,140 @@
+//===- ScheduleLinearPipeline.cpp - Linear pipeline scheduling pass -----*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Contains the definitions linear pipeline scheduling pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Analysis/DependenceAnalysis.h"
+#include "circt/Analysis/SchedulingAnalysis.h"
+#include "circt/Scheduling/Algorithms.h"
+#include "circt/Scheduling/Utilities.h"
+#include "circt/Support/BackedgeBuilder.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "pipeline-schedule-linear"
+
+using namespace mlir;
+using namespace circt;
+using namespace pipeline;
+
+namespace {
+
+class ScheduleLinearPipelinePass
+    : public ScheduleLinearPipelineBase<ScheduleLinearPipelinePass> {
+public:
+  void runOnOperation() override;
+};
+
+} // end anonymous namespace
+
+// Returns true if 'op' should be ignored in the scheduling problem.
+static bool ignoreOp(Operation *op) {
+  return isa<hw::ConstantOp, pipeline::ReturnOp>(op);
```

Comment:
Does it make sense to use the traits here? e.g. ConstantLike, ReturnLike?

Reply:
I like that - will change!